### PR TITLE
Add quick settings tiles for scripts and scenes

### PR DESCRIPTION
**Pull request by dshokouhi**
## Summary

Adds support for quick settings tiles to execute a script/scene!

https://developer.android.com/reference/android/service/quicksettings/package-summary

As we have to define each tile in the manifest and link it to its own class I have opted to start with 5 and based on user feedback we can add more easily.

Users are required to set the label and select an entity ID. Subtitle is optional and an icon can be selected too.  Once a tile is added we will update the tile data when it is visible and when the user clicks on the tile.  Adding a tile with proper data set will show as inactive and will be active once selected and will go back to inactive upon success.  Failures will show a toast message and tiles added without any data will be in an unavailable state.

## Screenshots

image

image

image

image

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#493

### CODE REVIEW

**Comment by JBassett** [comment]
Need to fix merge conflicts.

File: app/src/main/java/io/homeassistant/companion/android/settings/qs/ManageTilesFragment.kt [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                        entityList = entityList + it.entityId
+                }
+            } catch (e: Exception) {
+                Log.e(TAG, "Unable to fetch list of entity IDs", e)
+            }
+        }
+
+        if (entityList.isNotEmpty()) {
+            tileEntityPref?.entries = entityList.sorted().toTypedArray()
+            tileEntityPref?.entryValues = entityList.sorted().toTypedArray()
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/settings/qs/ManageTilesFragment.kt:117]: This is sorting and creating the same array 2 times.

**Review by JBassett — CHANGES_REQUESTED** [review]

File: app/src/main/java/io/homeassistant/companion/android/qs/Tile1Service.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import android.os.Build
+import android.service.quicksettings.TileService
+import androidx.annotation.RequiresApi
+import io.homeassistant.companion.android.common.dagger.GraphComponentAccessor
+import io.homeassistant.companion.android.common.data.integration.IntegrationRepository
+import javax.inject.Inject
+
+@RequiresApi(Build.VERSION_CODES.N)
+class Tile1Service : TileService() {
```
[the remark this answers is not in the document]
Comment (dshokouhi) [reply app/src/main/java/io/homeassistant/companion/android/qs/Tile1Service.kt]: Yea that is much cleaner, think I got it the way you mentioned :)

File: app/src/main/java/io/homeassistant/companion/android/qs/Tile4Service.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    @Inject
+    lateinit var integrationUseCase: IntegrationRepository
+    companion object {
+        private const val TILE_ID = "tile_4"
+    }
+    override fun onClick() {
+        super.onClick()
+        TileExtensions.setTileData(applicationContext, TILE_ID, qsTile)
+        TileExtensions.tileClicked(applicationContext, TILE_ID, qsTile, integrationUseCase)
+    }
```
[the remark this answers is not in the document]
Comment (dshokouhi) [reply app/src/main/java/io/homeassistant/companion/android/qs/Tile4Service.kt]: oops missed one :) luckily new logic only needs it in one place

File: app/src/main/java/io/homeassistant/companion/android/settings/qs/ManageTilesFragment.kt (hunk 1 above)
Comment (dshokouhi) [reply app/src/main/java/io/homeassistant/companion/android/settings/qs/ManageTilesFragment.kt]: ah good catch

File: app/src/main/java/io/homeassistant/companion/android/qs/TileExtensions.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import com.maltaisn.icondialog.pack.IconPack
+import com.maltaisn.icondialog.pack.IconPackLoader
+import com.maltaisn.iconpack.mdi.createMaterialDesignIconPack
+import io.homeassistant.companion.android.R
+import io.homeassistant.companion.android.common.data.integration.IntegrationRepository
+import io.homeassistant.companion.android.database.AppDatabase
+import kotlinx.coroutines.runBlocking
+
+@RequiresApi(Build.VERSION_CODES.N)
+class TileExtensions {
```
[the remark this answers is not in the document]
Comment (dshokouhi) [reply app/src/main/java/io/homeassistant/companion/android/qs/TileExtensions.kt]: yup agreed much better

**Review by JBassett — APPROVED** [review]

**Pull request merged**